ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotSelectorFromBlock.java

```diff
@@ -0,0 +1,166 @@
+      final EthContext ethContext,
+      final MetricsSystem metricsSystem,
+      final GenesisConfigOptions genesisConfig,
+      final Supplier<Optional<ForkchoiceEvent>> forkchoiceStateSupplier,
+      final Runnable cleanupAction) {
+    this.protocolContext = protocolContext;
+    this.protocolSchedule = protocolSchedule;
+    this.ethContext = ethContext;
+    this.metricsSystem = metricsSystem;
+    this.genesisConfig = genesisConfig;
+    this.forkchoiceStateSupplier = forkchoiceStateSupplier;
+    this.cleanupAction = cleanupAction;
+  }
+
+  @Override
+  public CompletableFuture<Void> prepareRetry() {
+    // nothing to do
+    return CompletableFuture.completedFuture(null);
+  }
+
+  @Override
+  public void close() {
+    cleanupAction.run();
+  }
+
+  @Override
+  public long getMinRequiredBlockNumber() {
+    return genesisConfig.getTerminalBlockNumber().orElse(0L);
+  }
+
+  @Override
+  public long getBestChainHeight() {
+    final long localChainHeight = protocolContext.getBlockchain().getChainHeadBlockNumber();
+
+    return Math.max(
+        forkchoiceStateSupplier
+            .get()
+            .map(ForkchoiceEvent::getHeadBlockHash)
+            .map(
+                headBlockHash ->
+                    maybeCachedHeadBlockHeader
+                        .filter(
+                            cachedBlockHeader -> cachedBlockHeader.getHash().equals(headBlockHash))
+                        .map(BlockHeader::getNumber)
+                        .orElseGet(
+                            () -> {
+                              LOG.debug(
+                                  "Downloading chain head block header by hash {}", headBlockHash);
+                              try {
+                                return waitForPeers(1)
+                                    .thenCompose(unused -> downloadBlockHeader(headBlockHash))
+                                    .thenApply(
+                                        blockHeader -> {
+                                          maybeCachedHeadBlockHeader = Optional.of(blockHeader);
+                                          return blockHeader.getNumber();
+                                        })
+                                    .get(20, TimeUnit.SECONDS);
+                              } catch (Throwable t) {
+                                LOG.debug(
+                                    "Error trying to download chain head block header by hash {}",
+                                    headBlockHash,
+                                    t);
+                              }
+                              return null;
+                            }))
+            .orElse(0L),
+        localChainHeight);
+  }
+
+  @Override
+  public Optional<FastSyncState> selectNewPivotBlock() {
+    final Optional<ForkchoiceEvent> maybeForkchoice = forkchoiceStateSupplier.get();
+    if (maybeForkchoice.isPresent() && maybeForkchoice.get().hasValidSafeBlockHash()) {
```

Comment:
hasValidSafeBlockHash() isn't necessary for the head block implementation.  I would suggest making getPivotHash return Optional<Hash>, and you can do the safe block check there and return Optional.empty() if it doesn't exist.

this is a very corner case, where we would be trying to sync a network very near genesis, before a block has been marked 'safe'.  The prior implementation of safe-only would have the same issue.  So this isn't a huge deal, but fixing it now might save a headache in the future for sync tests or something like that where we would be syncing a brand new chain.

Reply:
code removed